File: scripts/images/catalog.py
"""TOML 读取与跨文件聚合。"""

from __future__ import annotations

import tomllib
from dataclasses import dataclass, field
from pathlib import Path
from typing import TypeVar

from pydantic import BaseModel, TypeAdapter, ValidationError

from .models import (
    CHARACTER_POOLS,
    SECT_POOL,
    CharacterSource,
    EntityKind,
    Name,
    SectSource,
    SourceImage,
    ThemeSource,
)

SourceModel = TypeVar("SourceModel", bound=BaseModel)
NAME_ADAPTER = TypeAdapter(Name)


class BuildError(Exception):
    """包含源文件上下文的构建错误。"""


def _location(parts: tuple[str | int, ...]) -> str:
    return "$" + "".join(
        f"[{part}]" if isinstance(part, int) else f".{part}" for part in parts
    )


def _validation_details(error: ValidationError) -> str:
    return "; ".join(
        f"{_location(item['loc'])}: {item['msg']}"
        for item in error.errors(include_url=False)
    )


def load_source(path: Path, model: type[SourceModel]) -> SourceModel:
    try:
        with path.open("rb") as file:
            data = tomllib.load(file)
        return model.model_validate(data)
    except tomllib.TOMLDecodeError as error:
        raise BuildError(f"{path}: 非法 TOML: {error}") from error
    except ValidationError as error:
        raise BuildError(
            f"{path}: TOML 格式错误: {_validation_details(error)}"
        ) from error


def source_name(path: Path) -> str:
    try:
        return NAME_ADAPTER.validate_python(path.stem)
    except ValidationError as error:
        raise BuildError(f"{path}: 非法文件名: {_validation_details(error)}") from error


def _first_duplicate(values: list[str]) -> str | None:
    seen: set[str] = set()
    for value in values:
        if value in seen:
            return value
        seen.add(value)
    return None


def source_files(root: Path, directory: str) -> list[Path]:
    source_dir = root / "data" / directory
    if not source_dir.is_dir():
        raise BuildError(f"缺少数据目录: {source_dir}")
    return sorted(source_dir.glob("*.toml"))


@dataclass
class CatalogEntity:
    kind: EntityKind
    pools: dict[str, tuple[SourceImage, ...]] = field(default_factory=dict)
    origins: dict[str, Path] = field(default_factory=dict)


@dataclass
class Catalog:
    entities: dict[str, CatalogEntity] = field(default_factory=dict)
    themes: list[str] = field(default_factory=list)

    def add(
        self,
        *,
        name: str,
        kind: EntityKind,
        pool: str,
        images: list[SourceImage],
        source: Path,
    ) -> None:
        duplicate = _first_duplicate([image.url for image in images])
        if duplicate:
            raise BuildError(
                f"{source}: 实体 [{name}] 的池 [{pool}] 存在重复 URL: {duplicate}"
            )

        entity = self.entities.setdefault(name, CatalogEntity(kind=kind))
        if entity.kind != kind:
            raise BuildError(
                f"实体 [{name}] 被同时声明为 {entity.kind} 与 {kind}，命名冲突"
            )
        if previous := entity.origins.get(pool):
            raise BuildError(
                f"{source} 与 {previous} 重复定义了 [{name}] 的池 [{pool}]"
            )
        entity.pools[pool] = tuple(images)
        entity.origins[pool] = source


def load_catalog(root: Path) -> Catalog:
    catalog = Catalog()

    for path in source_files(root, "character"):
        source = load_source(path, CharacterSource)
        name = source_name(path)
        for pool, images in source.pools():
            catalog.add(
                name=name,
                kind="character",
                pool=pool,
                images=images,
                source=path,
            )

    for path in source_files(root, "sect"):
        source = load_source(path, SectSource)
        catalog.add(
            name=source_name(path),
            kind="sect",
            pool=SECT_POOL,
            images=source.images["map"],
            source=path,
        )

    for path in source_files(root, "themes"):
        theme = source_name(path)
        if theme in CHARACTER_POOLS:
            raise BuildError(f"{path}: 主题名 [{theme}] 与固定图片分类重名")
        source = load_source(path, ThemeSource)
        catalog.themes.append(theme)
        if not source.characters:
            continue
        for name, character in source.characters.items():
            images = [
                image.resolve(comment=source.comment, tags=source.tags)
                for image in character["images"]
            ]
            catalog.add(
                name=name,
                kind="character",
                pool=theme,
                images=images,
                source=path,
            )

    if not catalog.entities:
        raise BuildError("未找到任何图片数据")
    return catalog

File: src/cli/catalog.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { parse as parseToml } from "smol-toml";
import * as z from "zod";

import {
  CharacterSourceSchema,
  SectSourceSchema,
  ThemeSourceSchema,
  type SourceImage,
  CHARACTER_POOLS,
  SECT_POOL,
} from "./source-schema.js";
import { NameSchema } from "../sdk/schema.js";
import type { EntityKind } from "../sdk/types.js";

/** 包含源文件上下文的构建错误。 */
export class BuildError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "BuildError";
  }
}

export interface CatalogEntity {
  kind: EntityKind;
  pools: Record<string, SourceImage[]>;
  origins: Record<string, string>;
}

export interface Catalog {
  entities: Map<string, CatalogEntity>;
  themes: string[];
}

/** 按 Unicode 码点做字典序比较，与 Python `sorted` 对字符串的行为一致。 */
export function compareCodePoints(a: string, b: string): number {
  const left = Array.from(a);
  const right = Array.from(b);
  const length = Math.min(left.length, right.length);
  for (let i = 0; i < length; i += 1) {
    const lc = left[i]!.codePointAt(0)!;
    const rc = right[i]!.codePointAt(0)!;
    if (lc !== rc) return lc - rc;
  }
  return left.length - right.length;
}

function zodDetails(error: z.ZodError): string {
  return z.prettifyError(error);
}

function sourceFiles(root: string, directory: string): string[] {
  const sourceDir = path.join(root, "data", directory);
  if (!fs.existsSync(sourceDir) || !fs.statSync(sourceDir).isDirectory()) {
    throw new BuildError(`缺少数据目录: ${sourceDir}`);
  }
  return fs
    .readdirSync(sourceDir)
    .filter((name) => name.endsWith(".toml"))
    .sort(compareCodePoints)
    .map((name) => path.join(sourceDir, name));
}

function loadSource<Schema extends z.ZodType>(
  filePath: string,
  schema: Schema,
): z.output<Schema> {
  let data: unknown;
  try {
    data = parseToml(fs.readFileSync(filePath, "utf8"));
  } catch (error) {
    throw new BuildError(`${filePath}: 非法 TOML: ${(error as Error).message}`);
  }
  const result = schema.safeParse(data);
  if (!result.success) {
    throw new BuildError(`${filePath}: TOML 格式错误: ${zodDetails(result.error)}`);
  }
  return result.data;
}

function sourceName(filePath: string): string {
  const stem = path.basename(filePath, ".toml");
  const result = NameSchema.safeParse(stem);
  if (!result.success) {
    throw new BuildError(`${filePath}: 非法文件名: ${zodDetails(result.error)}`);
  }
  return result.data;
}

function firstDuplicate(values: string[]): string | undefined {
  const seen = new Set<string>();
  for (const value of values) {
    if (seen.has(value)) return value;
    seen.add(value);
  }
  return undefined;
}

/** 应用主题根级默认值：comment 缺省回退；tags 未声明回退根级、显式空退出、非空并集去重（根级在前）。 */
function resolveImage(
  source: SourceImage,
  comment: string | undefined,
  tags: string[],
): SourceImage {
  let mergedTags: string[];
  if (source.tags === undefined) {
    mergedTags = [...tags];
  } else if (source.tags.length > 0) {
    mergedTags = [...new Set([...tags, ...source.tags])];
  } else {
    mergedTags = [];
  }
  return {
    url: source.url,
    comment: source.comment !== undefined ? source.comment : comment,
    tags: mergedTags,
  };
}

interface AddOptions {
  name: string;
  kind: EntityKind;
  pool: string;
  images: SourceImage[];
  source: string;
}

function add(catalog: Catalog, options: AddOptions): void {
  const { name, kind, pool, images, source } = options;
  const duplicate = firstDuplicate(images.map((image) => image.url));
  if (duplicate) {
    throw new BuildError(`${source}: 实体 [${name}] 的池 [${pool}] 存在重复 URL: ${duplicate}`);
  }

  let entity = catalog.entities.get(name);
  if (!entity) {
    entity = { kind, pools: {}, origins: {} };
    catalog.entities.set(name, entity);
  }
  if (entity.kind !== kind) {
    throw new BuildError(`实体 [${name}] 被同时声明为 ${entity.kind} 与 ${kind}，命名冲突`);
  }
  const previous = entity.origins[pool];
  if (previous) {
    throw new BuildError(`${source} 与 ${previous} 重复定义了 [${name}] 的池 [${pool}]`);
  }
  entity.pools[pool] = images;
  entity.origins[pool] = source;
}

export function loadCatalog(root: string): Catalog {
  const catalog: Catalog = { entities: new Map(), themes: [] };

  for (const filePath of sourceFiles(root, "character")) {
    const source = loadSource(filePath, CharacterSourceSchema);
    const name = sourceName(filePath);
    for (const pool of CHARACTER_POOLS) {
      const images = source.images[pool];
      if (images) {
        add(catalog, { name, kind: "character", pool, images, source: filePath });
      }
    }
  }

  for (const filePath of sourceFiles(root, "sect")) {
    const source = loadSource(filePath, SectSourceSchema);
    add(catalog, {
      name: sourceName(filePath),
      kind: "sect",
      pool: SECT_POOL,
      images: source.images.map,
      source: filePath,
    });
  }

  for (const filePath of sourceFiles(root, "themes")) {
    const theme = sourceName(filePath);
    if ((CHARACTER_POOLS as readonly string[]).includes(theme)) {
      throw new BuildError(`${filePath}: 主题名 [${theme}] 与固定图片分类重名`);
    }
    const source = loadSource(filePath, ThemeSourceSchema);
    catalog.themes.push(theme);
    if (!source.characters) continue;
    for (const [name, character] of Object.entries(source.characters)) {
      const images = character.images.map((image) =>
        resolveImage(image, source.comment, source.tags ?? []),
      );
      add(catalog, { name, kind: "character", pool: theme, images, source: filePath });
    }
  }

  if (catalog.entities.size === 0) {
    throw new BuildError("未找到任何图片数据");
  }
  return catalog;
}
